include/deal.II/lac/scalapack.h

```diff
@@ -180,21 +180,84 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
    */
   void invert();
 
-  /**
-   * Compute all eigenvalues of a real symmetric matrix using <code>pXsyev</code>.
-   * If successful, the computed eigenvalues are arranged in ascending order.
-   * After this function is called, the content of the matrix is overwritten
-   * making it unusable.
-   */
-  std::vector<NumberType> eigenvalues_symmetric();
+
 
   /**
-   * Compute all eigenpairs of a real symmetric matrix using <code>pXsyev</code>.
+   * Function to compute selected eigenvalues and, optionally, the eigenvectors.
+   * If the function is called with the default arguments all eigenvalues are computed but no eigenvectors.
+   * The eigenvalues/eigenvectors are selected by either prescribing a range of indices @p index_limits
+   * or a range of values @p value_limits for the eigenvalues. The funtion will throw an exception
+   * if both ranges are prescribed (meaning that both ranges differ from the default value)
+   * as this ambiguity is prohibited.
    * If successful, the computed eigenvalues are arranged in ascending order.
    * The eigenvectors are stored in the columns of the matrix, thereby
    * overwriting the original content of the matrix.
    */
-  std::vector<NumberType> eigenpairs_symmetric ();
+  std::vector<NumberType> eigenpairs_symmetric(const bool compute_eigenvectors=false,
+                                               const std::pair<int,int> &index_limits = std::make_pair(-1,-1),
+                                               const std::pair<NumberType,NumberType> &value_limits = std::make_pair(-1,-1));
+
+
+
+  /**
+  * Funcion to compute the singular value decomposition (SVD) of an
+  * M-by-N matrix A, optionally computing the left and/or right
+  * singular vectors. The SVD is written as A = U * SIGMA * transpose(V)
+  * where SIGMA is an M-by-N diagonal matrix, @p U is an M-by-M orthogonal matrix,
+  * and @p V is an N-by-N orthogonal matrix. The diagonal elements of SIGMA
+  * are the singular values of A and the columns of U and V are the
+  * corresponding left and right singular vectors, respectively. The
+  * singular values are returned in decreasing order and only the first min(M,N)
+  * columns of U and rows of VT = transpose(V) are computed.
+  * Upon return the content of the matrix is unusable.
+  * The matrix A must have identical block cyclic distribution for the rows and column
+  * If left singular vectors are required matrices A and U
+  * have to be constructed with the same process grid and block cyclic distribution.
+  * If right singular vectors are required matrices A and VT
+  * have to be constructed with the same process grid  and block cyclic distribution.
+  */
+  std::vector<NumberType> compute_SVD(ScaLAPACKMatrix<NumberType> &U,
```

Comment:
I think that would be a perfect place for `boost::optional<>`. @dealii/dealii thoughts?

Reply:
Ps. Don’t rush with this one, I am not sure how boost::optional will work with references. Let’s wait for opinions on this

Reply:
@davydden -- can you clarify what you want `boost::optional` to be used for? Is it because the user has to provide *either* `U` or `VT` here, and you want to use `boost::optional` to allow not specifying anything that's not used?

Reply:
correct, my understanding that both matrices are optional.

One could of course get the same effect with two pointers, with default being nullptr. Perhaps that’s a better way

Reply:
Yes, maybe.

Reply:
Function will receive pointers to both variables and only perform the calculations if the pointers are not nullptr.